### ISSUE TITLE
Master can not ping pods, request disappears after PREROUTING chain of nat iptable

### ISSUE THREAD

**Issue by MikeSpreitzer** [projectcalico/calicoctl#1377]
Ubuntu 16.04
Kubernetes release 1.4.6
Calico 1.6 (calicoctl and calico/node v0.23.0)
Calico CNI 1.4.3

On the worker nodes, I have a calico-node service with
```
ExecStart=/usr/bin/docker run --net=host --privileged --name=calico-node \
    -e IP=100.126.0.5 \
    -e NO_DEFAULT_POOLS=true \
    -e CALICO_LIBNETWORK_ENABLED=true \
    -e ETCD_ENDPOINTS=https://100.126.0.2:4001 \
     \
    -e ETCD_CA_CERT_FILE=/etc/kubernetes/cert/ca.pem \
    -e ETCD_CERT_FILE=/etc/kubernetes/cert/etcd.pem \
    -e ETCD_KEY_FILE=/etc/kubernetes/cert/etcd-key.pem \
    -v /etc/kubernetes/cert:/etc/kubernetes/cert \
     \
    -v /var/log/calico:/var/log/calico \
    -v /run/docker/plugins:/run/docker/plugins \
    -v /lib/modules:/lib/modules \
    -v /var/run/calico:/var/run/calico \
    calico/node:v0.23.0
```

After starting the calico-node service on the master and all the workers, my Ansible playbook then creates one pool; here is what was logged in the typescript:
```
changed: [100.126.0.3] => {"changed": true, "cmd": "ETCD_ENDPOINTS=https://100.126.0.2:4001 ETCD_CA_CERT_FILE=/etc/kubernetes/cert/ca.pem ETCD_CERT_FILE=/etc/kubernetes/cert/etcd.pem ETCD_KEY_FILE=/etc/kubernetes/cert/etcd-key.pem /opt/bin/calicoctl pool add 172.18.0.0/16  --ipip   --nat-outgoing ", "delta": "0:00:00.533429", "end": "2016-12-01 14:03:28.176998", "rc": 0, "start": "2016-12-01 14:03:27.643569", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}
```

Later the playbook goes on to install Kubernetes components, and then runs the Calico policy controller in Kubernetes.

When all the setup is done I have a working system except for one thing: the master can not send packets to the pods.  `ip route` is fine everywhere.  When the master tries to ping a pod, `tcpdump -i tunl0` on the worker shows the ping request coming out of the tunnel, but `tcpdump -i calibe9b1a97acc` shows that the packet does not get to the network interface (in the worker node's main network namespace) that leads to the pod.

I investigated further by adding `-j LOG` rules in certain iptables chains (ones that felix will not fight over) on the worker node hosting the pod. Doing so I see the ping request logged in the last rule of the PREROUTING chain of the `nat` table.  But the packet is not logged in the FORWARD chain --- nor the INPUT chain --- of the `mangle` table.  I do not understand how this is even possible.

To deepen the mystery, if another worker node tries to ping the same pod, then that succeeds (and the iptables LOG output shows the proper progress for the packet).

If I then use `calicoctl pool add 172.18.0.0/16 --nat-outgoing` to remove the ip-in-ip setting, and then `calicoctl pool add 172.18.0.0/16 --ipip --nat-outgoing` to add it back, then the master can ping the pod (and the iptables logging shows the packet making proper progress).

**Comment by MikeSpreitzer** [comment]
Here is `calicoctl diags`, taken at this point in the narrative (i.e., after the problem was made to evaporate by removing and re-adding ipip):
[diags-021216_004457.tar.gz](https://github.com/projectcalico/calico-containers/files/626451/diags-021216_004457.tar.gz)

**Comment by MikeSpreitzer** [comment]
Here is a dump of the iptables on the worker node from a certain point in time while I was probing the bad behavior.

[iptables-with-loggin-and-problem.txt](https://github.com/projectcalico/calico-containers/files/626458/iptables-with-loggin-and-problem.txt)

This was after some `-j LOG` rules had been added but not all.  After this point I added three more:

```
root@mvespa2-5:/var/log# iptables -t mangle -A FORWARD -j LOG --log-prefix "[MJS man FWD] " --log-tcp-options --log-ip-options

root@mvespa2-5:/var/log# iptables -t mangle -A POSTROUTING -j LOG --log-prefix "[MJS man PSR] " --log-tcp-options --log-ip-options

root@mvespa2-5:/var/log# iptables -t mangle -A INPUT -j LOG --log-prefix "[MJS man INP] " --log-tcp-options --log-ip-options
```

**Comment by MikeSpreitzer** [comment]
To complete your viewing pleasure, here is the output of `fgrep '[MJS' syslog` at 23:28 on the worker node's clock.
[syslog-filtered.zip](https://github.com/projectcalico/calico-containers/files/626461/syslog-filtered.zip)

**Comment by MikeSpreitzer** [comment]
Finally, for easy reference:
the master was at 100.126.0.2 and Calico assigned 172.18.118.0/26;
the worker node hosting the pod was at 100.126.0.5 and Calico assigned 172.18.157.128/26;
I successfully pinged from a worker node at 100.126.0.3, where Calico assigned 172.18.87.128/26.

Interestingly, this is one of the times when I found Calico assigned two blocks to each host.  Here is the routing table on the master, at this final point in the narrative:
```
$ ip route
default via 100.126.0.1 dev enp0s3 
100.126.0.0/24 dev enp0s3  proto kernel  scope link  src 100.126.0.2 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown 
172.18.39.64/26 via 100.126.0.6 dev tunl0  proto bird onlink 
172.18.87.128/26 via 100.126.0.3 dev tunl0  proto bird onlink 
172.18.105.192/26 via 100.126.0.5 dev tunl0  proto bird onlink 
blackhole 172.18.118.0/26  proto bird 
172.18.122.192/26 via 100.126.0.4 dev tunl0  proto bird onlink 
172.18.149.128/26 via 100.126.0.6 dev tunl0  proto bird onlink 
172.18.152.128/26 via 100.126.0.4 dev tunl0  proto bird onlink 
172.18.157.128/26 via 100.126.0.5 dev tunl0  proto bird onlink 
172.18.255.192/26 via 100.126.0.3 dev tunl0  proto bird onlink 
```

And here is the network interface list at the end of the narrative:
```
$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1446 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:38:02:02 brd ff:ff:ff:ff:ff:ff
    inet 100.126.0.2/24 brd 100.126.0.255 scope global enp0s3
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe38:202/64 scope link 
       valid_lft forever preferred_lft forever
3: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 02:42:99:13:6c:b7 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:99ff:fe13:6cb7/64 scope link 
       valid_lft forever preferred_lft forever
6: tunl0@NONE: <NOARP,UP,LOWER_UP> mtu 1440 qdisc noqueue state UNKNOWN group default qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
    inet 172.18.118.1/32 scope global tunl0
       valid_lft forever preferred_lft forever
```

**Comment by MikeSpreitzer** [comment]
Here a clue I noticed in the iptables logging: when it was failing, the request packet's source address is 172.18.118.0; after the repair, the successful packet's source address is 172.18.118.1.

**Comment by MikeSpreitzer** [comment]
I changed my Ansible playbook to create the desired Calico pool before starting the calico-node service anywhere, and this works around the problem.

**Comment by fasaxc** [comment]
@MikeSpreitzer Did you happen to look at the routing table on the target machine before it fixed itself?  We've had some issues with BIRD where it doesn't correctly update the routes after IPIP is toggled on.  A correct IPIP route looks like this:
```
172.18.122.192/26 via 100.126.0.4 dev tunl0  proto bird onlink
```
i.e. 

- it matches the `/26` IPAM block
- has a via for the real, routable IP of the destination host
- it has an exlicit `dev tunl0` and `onlink` clause.

Another thing to check is that the source address for a host -> container packet needs to be the tunnel address (which is the IP chosen for a host from one of its /26 IPAM blocks) or RPF checks will fail or the return packet will fail to go via the tunnel.

We believe that BIRD issue has been fixed by revving BIRD in the upcoming v2.0.0 release of Calico.

That you're getting a second IPAM block allocated looks a bit fishy to me.  Did that correlate with toggling IPIP off and on again?

**Comment by MikeSpreitzer** [comment]
The double blocks was there from the start, the repair procedure did not introduce it.  Now that I have destroyed the troubled cluster and re-built it with the work-around, I also see double blocks.  Here is the routing table on the master:
```
root@mvespa2-2:~# ip route
default via 100.126.0.1 dev enp0s3 
100.126.0.0/24 dev enp0s3  proto kernel  scope link  src 100.126.0.2 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown 
172.18.87.128/26 via 100.126.0.3 dev tunl0  proto bird onlink 
172.18.90.128/26 via 100.126.0.4 dev tunl0  proto bird onlink 
172.18.105.192/26 via 100.126.0.5 dev tunl0  proto bird onlink 
blackhole 172.18.118.0/26  proto bird 
172.18.149.128/26 via 100.126.0.6 dev tunl0  proto bird onlink 
172.18.152.128/26 via 100.126.0.4 dev tunl0  proto bird onlink 
172.18.176.192/26 via 100.126.0.3 dev tunl0  proto bird onlink 
172.18.186.64/26 via 100.126.0.6 dev tunl0  proto bird onlink 
172.18.202.128/26 via 100.126.0.5 dev tunl0  proto bird onlink 
```

Yes, in the troubled case I looked at `ip route` on the master and on some workers, and did not see anything wrong.  Unfortunately I do not have a handy record of exactly what I saw.  If it is helpful I can create a cluster using the older Ansible playbook without the work-around.  But remember that the problem was not the routing of the packet from the sender to the target; I verified with `tcpdump -i tunl0` on the destination and with iptables `-j LOG` on the destination that the packet arrived there.

Yes, the packet arrived with a source address from an address block allocated by Calico.  My earlier remarks and some of the uploaded data files show that.  The problem is not that the return packet cannot get back.  The problem is that the request packet disappears in the midst of iptables processing.  You can see from the iptables logging I added exactly where it disappears.

**Comment by fasaxc** [comment]
> Yes, the packet arrived with a source address from an address block allocated by Calico

Yes, sorry you did confirm that in a previous comment.

My working hypothesis was that the packet was failing the kernel's RPF check on the destination host, which could happen if BIRD had failed to correct the routes after turning on IPIP.  I'm not 100% sure where in the pipeline the RPF check occurs but the routing decision is made between `nat PREROUTING` and `mangle FORWARD`, which is a pretty likely place.

If you did repro the scenario with the old playbook, then it'd help to have calico diags output from the broken state on source and destination hosts (or just reach out on slack and one of us should be able to help diagnose the live, broken system).

**Comment by MikeSpreitzer** [comment]
OK, I have reproduced the problem with the old playbook.  Same as before, Ubuntu 16.04, Calico 1.6, Kubernetes 1.4.6.
From the master (100.127.0.2 AKA 172.16.184.128), pinging a pod fails.  The ping request appears at `tunl0` on the destination worker node, but not on the `califbe0e613d1c` interface in that node's main network namespace (leading to the pod).  OTOH, a ping from another worker node (100.127.0.3 AKA 172.16.46.192) succeeds, with the ping request appearing on the destination worker node at both `tunl0` and `califbe0e613d1c`.

I ran two copies of `tcpdump` simultaneously on the destination worker node simultaneously, one looking at `tunl0` and one at `califbe0e613d1c`.  While that packet capture was running, I pinged first from the master and then from the other worker.  Then I displayed the captured packets, showing what I said above.  Here is the typescript from the capture and display from `tunl0`:
```
root@mvespa1-5:~# tcpdump -i tunl0 -w try2-host5-tunl0.pcap
tcpdump: listening on tunl0, link-type RAW (Raw IP), capture size 262144 bytes
^C5 packets captured
5 packets received by filter
0 packets dropped by kernel
root@mvespa1-5:~# 
root@mvespa1-5:~# tcpdump -r try2-host5-tunl0.pcap -enn
reading from file try2-host5-tunl0.pcap, link-type RAW (Raw IP)
15:27:49.465073 ip: 172.16.184.128 > 172.16.228.1: ICMP echo request, id 17592, seq 1, length 64
15:28:01.519894 ip: 172.16.46.192 > 172.16.228.1: ICMP echo request, id 24481, seq 1, length 64
15:28:01.519964 ip: 172.16.228.1 > 172.16.46.192: ICMP echo reply, id 24481, seq 1, length 64
15:28:02.518931 ip: 172.16.46.192 > 172.16.228.1: ICMP echo request, id 24481, seq 2, length 64
15:28:02.518972 ip: 172.16.228.1 > 172.16.46.192: ICMP echo reply, id 24481, seq 2, length 64
```

Here is the typescript of the capture and display for `califbe0e613d1c`:
```
root@mvespa1-5:~# tcpdump -i califbe0e613d1c -w try2-host5-cali.pcap
tcpdump: listening on califbe0e613d1c, link-type EN10MB (Ethernet), capture size 262144 bytes
^C4 packets captured
4 packets received by filter
0 packets dropped by kernel
root@mvespa1-5:~# 
root@mvespa1-5:~# 
root@mvespa1-5:~# tcpdump -r try2-host5-cali.pcap -enn
reading from file try2-host5-cali.pcap, link-type EN10MB (Ethernet)
15:28:01.519930 a6:2c:c4:fd:b6:ca > 52:7a:0c:13:a2:9f, ethertype IPv4 (0x0800), length 98: 172.16.46.192 > 172.16.228.1: ICMP echo request, id 24481, seq 1, length 64
15:28:01.519958 52:7a:0c:13:a2:9f > a6:2c:c4:fd:b6:ca, ethertype IPv4 (0x0800), length 98: 172.16.228.1 > 172.16.46.192: ICMP echo reply, id 24481, seq 1, length 64
15:28:02.518945 a6:2c:c4:fd:b6:ca > 52:7a:0c:13:a2:9f, ethertype IPv4 (0x0800), length 98: 172.16.46.192 > 172.16.228.1: ICMP echo request, id 24481, seq 2, length 64
15:28:02.518967 52:7a:0c:13:a2:9f > a6:2c:c4:fd:b6:ca, ethertype IPv4 (0x0800), length 98: 172.16.228.1 > 172.16.46.192: ICMP echo reply, id 24481, seq 2, length 64
```

Here is the typescript of the ping from the master:
```
root@mvespa1-2:~# date; ping -c 1 172.16.228.1; date
Fri Dec  2 15:27:49 EST 2016
PING 172.16.228.1 (172.16.228.1) 56(84) bytes of data.

--- 172.16.228.1 ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms

Fri Dec  2 15:27:59 EST 2016
```

Here is the typescript of the ping from the other worker:
```
mspreitz@mvespa1-3:~$ date; ping -c 2 172.16.228.1; date
Fri Dec  2 15:28:01 EST 2016
PING 172.16.228.1 (172.16.228.1) 56(84) bytes of data.
64 bytes from 172.16.228.1: icmp_seq=1 ttl=63 time=0.697 ms
64 bytes from 172.16.228.1: icmp_seq=2 ttl=63 time=0.826 ms

--- 172.16.228.1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 999ms
rtt min/avg/max/mdev = 0.697/0.761/0.826/0.070 ms
Fri Dec  2 15:28:02 EST 2016
```

**Comment by MikeSpreitzer** [comment]
I used `calicoctl diags` on the destination worker machine, and it partially succeeded.
```
root@mvespa1-5:~# calicoctl diags
Collecting diagnostics
Using temp dir: /tmp/tmp8AdcK9
Dumping netstat
Dumping routes (IPv4)
Dumping routes (IPv6)
Dumping interface info (IPv4)
Dumping interface info (IPv6)
Dumping iptables (IPv4)
Dumping iptables (IPv6)
Dumping ipsets
Copying journal for calico-node.service
Dumping felix stats
Copying Calico logs
Dumping datastore
Unable to dump etcd datastore

Diags saved to /tmp/tmp8AdcK9/diags-021216_154239.tar.gz

If required, you can upload the diagnostics bundle to a file sharing service
such as transfer.sh using curl or similar.  For example:

  curl --upload-file /tmp/tmp8AdcK9/diags-021216_154239.tar.gz https://transfer.sh/diags-021216_154239.tar.gz
```
[diags-021216_154239.tar.gz](https://github.com/projectcalico/calico-containers/files/628189/diags-021216_154239.tar.gz)

**Comment by MikeSpreitzer** [comment]
Here are the `calicoctl diags` archives from the master
[diags-021216_155225.tar.gz](https://github.com/projectcalico/calico-containers/files/628205/diags-021216_155225.tar.gz)
and the destination worker
[diags-021216_155208.tar.gz](https://github.com/projectcalico/calico-containers/files/628206/diags-021216_155208.tar.gz)

**Comment by MikeSpreitzer** [comment]
Yes, it looks like RPF is involved.  In my cluster with the problem, here is the routing table on a worker node running a pod that the master can not ping:
```
root@mvespa1-5:~# ip route
default via 100.127.0.1 dev enp0s3 
100.127.0.0/24 dev enp0s3  proto kernel  scope link  src 100.127.0.5 
blackhole 172.16.26.128/26  proto bird 
172.16.46.192/26 via 100.127.0.3 dev tunl0  proto bird onlink 
172.16.78.0/26 via 100.127.0.3 dev tunl0  proto bird onlink 
172.16.78.64/26 via 100.127.0.4 dev tunl0  proto bird onlink 
172.16.114.0/26 via 100.127.0.6 dev tunl0  proto bird onlink 
172.16.178.64/26 via 100.127.0.6 dev tunl0  proto bird onlink 
172.16.184.128/26 via 100.127.0.2 dev enp0s3  proto bird 
172.16.185.64/26 via 100.127.0.4 dev tunl0  proto bird onlink 
172.16.228.0 dev calid4eeb15d290  scope link 
blackhole 172.16.228.0/26  proto bird 
172.16.228.1 dev califbe0e613d1c  scope link 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown 
```

.. and here is what I see there coming in from the tunnel:
```
root@mvespa1-5:~# tcpdump -i tunl0 -enn
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on tunl0, link-type RAW (Raw IP), capture size 262144 bytes
22:21:54.234300 ip: 172.16.184.128 > 172.16.228.1: ICMP echo request, id 10087, seq 1, length 64
22:21:55.235796 ip: 172.16.184.128 > 172.16.228.1: ICMP echo request, id 10087, seq 2, length 64
22:21:56.235627 ip: 172.16.184.128 > 172.16.228.1: ICMP echo request, id 10087, seq 3, length 64
```

Note that the routing table entry for the request packets' source address is
```
172.16.184.128/26 via 100.127.0.2 dev enp0s3  proto bird
```

On the other hand, in the cluster built with the work-around, here is what I see arriving at a worker at the tunnel interface:
```
root@mvespa2-4:~# tcpdump -i tunl0 -enn
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on tunl0, link-type RAW (Raw IP), capture size 262144 bytes
22:23:49.743588 ip: 172.18.118.0 > 172.18.90.128: ICMP echo request, id 11626, seq 1, length 64
22:23:49.743933 ip: 172.18.90.128 > 172.18.118.0: ICMP echo reply, id 11626, seq 1, length 64
22:23:50.746805 ip: 172.18.118.0 > 172.18.90.128: ICMP echo request, id 11626, seq 2, length 64
22:23:50.746925 ip: 172.18.90.128 > 172.18.118.0: ICMP echo reply, id 11626, seq 2, length 64
```

.. and here is the routing table:
```
root@mvespa2-4:~# ip route
default via 100.126.0.1 dev enp0s3 
100.126.0.0/24 dev enp0s3  proto kernel  scope link  src 100.126.0.4 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown 
172.18.87.128/26 via 100.126.0.3 dev tunl0  proto bird onlink 
172.18.90.128 dev calibda3ef589a0  scope link 
blackhole 172.18.90.128/26  proto bird 
172.18.118.0/26 via 100.126.0.2 dev tunl0  proto bird onlink 
172.18.149.128/26 via 100.126.0.6 dev tunl0  proto bird onlink 
blackhole 172.18.152.128/26  proto bird 
172.18.176.192/26 via 100.126.0.3 dev tunl0  proto bird onlink 
172.18.186.64/26 via 100.126.0.6 dev tunl0  proto bird onlink 
```

**Comment by fasaxc** [comment]
CC @lxpollitt

**Comment by matthewdupre** [comment]
Sorry I haven't read all the way through; but is this similar to https://github.com/projectcalico/calicoctl/issues/1296?

**Comment by MikeSpreitzer** [comment]
This is different from #1296.

**Comment by Huang-Wei** [comment]
I'm afraid I got into exactly the same issue:

**Env:**

- Ubuntu 16.04
- Kubernetes release 1.4.5
- Calico 2.0.1 (calicoctl and calico/node v1.0.1)
- Calico CNI 1.5.5
- IPIP enabled

**Symptoms:**

The cluster has been running for a while, and today I found that one pod in one worker can't be pinged from any other workers/masters. And weird thing is another pod on the same worker could be pinged.

On the problematic worker, the routes for working and not-working pods:

```
$ ip r | grep 172.16.224
blackhole 172.16.224.128/26  proto bird
172.16.224.138 dev cali130823037c5  scope link
172.16.224.145 dev cali62fd3d1a193  scope link
172.16.224.146 dev cali2037eb03159  scope link
172.16.224.147 dev calidfcb97971e1  scope link
172.16.224.154 dev cali9c1176044b8  scope link
172.16.224.155 dev cali62fdab50ff3  scope link
172.16.224.156 dev cali66f1694f020  scope link
172.16.224.157 dev cali9d070f34956  scope link
172.16.224.162 dev cali9583e10ead8  scope link
172.16.224.168 dev cali20bf7f895dd  scope link
172.16.224.171 dev calib4427696e99  scope link
172.16.224.187 dev cali32900045d40  scope link
```

From any other worker/master, ping to `172.16.224.157` succeeded, but no luck to `172.16.224.147`.

```
$ ping -c1 172.16.224.157
PING 172.16.224.157 (172.16.224.157) 56(84) bytes of data.
64 bytes from 172.16.224.157: icmp_seq=1 ttl=63 time=0.447 ms
$ ping -c1 172.16.224.147
PING 172.16.224.147 (172.16.224.147) 56(84) bytes of data.
^C
```

**Analysis:**

Similar like @MikeSpreitzer, I dump the iptables logs by:

```
iptables -t mangle -A PREROUTING -p icmp -j LOG --log-prefix "ICMP (mangle.PREROUTING): "
iptables -t nat -A PREROUTING -p icmp -j LOG --log-prefix "ICMP (nat.PREROUTING): "
iptables -t mangle -A INPUT -p icmp -j LOG --log-prefix "ICMP (mangle.INPUT): "
iptables -t filter -A INPUT -p icmp -j LOG --log-prefix "ICMP (filter.INPUT): "
iptables -t mangle -A FORWARD -p icmp -j LOG --log-prefix "ICMP (mangle.FORWARD): "
iptables -t filter -A FORWARD -p icmp -j LOG --log-prefix "ICMP (filter.FORWARD): "

iptables -t mangle -A OUTPUT -p icmp -j LOG --log-prefix "ICMP (mangle.OUTPUT): "
iptables -t nat -A OUTPUT -p icmp -j LOG --log-prefix "ICMP (nat.OUTPUT): "
iptables -t filter -A OUTPUT -p icmp -j LOG --log-prefix "ICMP (filter.OUTPUT): "
iptables -t mangle -A POSTROUTING -p icmp -j LOG --log-prefix "ICMP (mangle.POSTROUTING): "
iptables -t nat -A POSTROUTING -p icmp -j LOG --log-prefix "ICMP (nat.POSTROUTING): "
```

For a successful ping (ping -c1 172.16.224.157), iptables logs looks like this:

```
Mar  7 09:00:24 problematic-worker kernel: [600803.046279] ICMP (mangle.PREROUTING): IN=tunl0 OUT= MAC= SRC=172.16.166.0 DST=172.16.224.157 LEN=84 TOS=0x00 PREC=0x00 T
TL=64 ID=25643 DF PROTO=ICMP TYPE=8 CODE=0 ID=20555 SEQ=1
Mar  7 09:00:24 problematic-worker kernel: [600803.046309] ICMP (nat.PREROUTING): IN=tunl0 OUT= MAC= SRC=172.16.166.0 DST=172.16.224.157 LEN=84 TOS=0x00 PREC=0x00 TTL=
64 ID=25643 DF PROTO=ICMP TYPE=8 CODE=0 ID=20555 SEQ=1
Mar  7 09:00:24 problematic-worker kernel: [600803.046316] ICMP (mangle.FORWARD): IN=tunl0 OUT=cali9d070f34956 MAC= SRC=172.16.166.0 DST=172.16.224.157 LEN=84 TOS=0x00
 PREC=0x00 TTL=63 ID=25643 DF PROTO=ICMP TYPE=8 CODE=0 ID=20555 SEQ=1
Mar  7 09:00:24 problematic-worker kernel: [600803.046325] ICMP (mangle.POSTROUTING): IN= OUT=cali9d070f34956 SRC=172.16.166.0 DST=172.16.224.157 LEN=84 TOS=0x00 PREC=
0x00 TTL=63 ID=25643 DF PROTO=ICMP TYPE=8 CODE=0 ID=20555 SEQ=1 MARK=0x3000000
Mar  7 09:00:24 problematic-worker kernel: [600803.046334] ICMP (nat.POSTROUTING): IN= OUT=cali9d070f34956 SRC=172.16.166.0 DST=172.16.224.157 LEN=84 TOS=0x00 PREC=0x0
0 TTL=63 ID=25643 DF PROTO=ICMP TYPE=8 CODE=0 ID=20555 SEQ=1 MARK=0x3000000
Mar  7 09:00:24 problematic-worker kernel: [600803.046406] ICMP (mangle.PREROUTING): IN=cali9d070f34956 OUT= MAC=06:cc:4b:2e:df:57:da:9e:09:7e:d7:f2:08:00 SRC=172.16.2
24.157 DST=172.16.166.0 LEN=84 TOS=0x00 PREC=0x00 TTL=64 ID=41131 PROTO=ICMP TYPE=0 CODE=0 ID=20555 SEQ=1
Mar  7 09:00:24 problematic-worker kernel: [600803.046413] ICMP (mangle.FORWARD): IN=cali9d070f34956 OUT=tunl0 MAC=06:cc:4b:2e:df:57:da:9e:09:7e:d7:f2:08:00 SRC=172.16
.224.157 DST=172.16.166.0 LEN=84 TOS=0x00 PREC=0x00 TTL=63 ID=41131 PROTO=ICMP TYPE=0 CODE=0 ID=20555 SEQ=1
Mar  7 09:00:24 problematic-worker kernel: [600803.046417] ICMP (mangle.POSTROUTING): IN= OUT=tunl0 SRC=172.16.224.157 DST=172.16.166.0 LEN=84 TOS=0x00 PREC=0x00 TTL=6
3 ID=41131 PROTO=ICMP TYPE=0 CODE=0 ID=20555 SEQ=1
```

For a failed ping (ping -c1 172.16.224.147), iptables logs looks like this:

```
Mar  7 09:08:56 problematic-worker kernel: [601314.964625] ICMP (mangle.PREROUTING): IN=tunl0 OUT= MAC= SRC=172.16.166.0 DST=172.16.224.147 LEN=84 TOS=0x00 PREC=0x00 TTL=64 ID=20120 DF PROTO=ICMP TYPE=8 CODE=0 ID=38420 SEQ=1
Mar  7 09:08:56 problematic-worker kernel: [601314.964729] ICMP (nat.PREROUTING): IN=tunl0 OUT= MAC= SRC=172.16.166.0 DST=172.16.224.147 LEN=84 TOS=0x00 PREC=0x00 TTL=64 ID=20120 DF PROTO=ICMP TYPE=8 CODE=0 ID=38420 SEQ=1
Mar  7 09:08:56 problematic-worker kernel: [601314.964739] ICMP (mangle.FORWARD): IN=tunl0 OUT=calidfcb97971e1 MAC= SRC=172.16.166.0 DST=172.16.224.147 LEN=84 TOS=0x00 PREC=0x00 TTL=63 ID=20120 DF PROTO=ICMP TYPE=8 CODE=0 ID=38420 SEQ=1
```

For the failed ping, I also inserted a rule at number 1 (`-t filter -I FORWARD 1`) to test whether the packet went into chain `FORWARD` of table `filter`. And the result implied NO.

So it's very likely the packet was totally dropped after the last rule of chain `FORWARD` of table `mangle`.

And thoughts?

**Comment by fasaxc** [comment]
@Huang-Wei Can you ping the workload from its own host?

**Comment by Huang-Wei** [comment]
@fasaxc I can ping from its own host.

For the incoming packet (from ip `10.177.83.90` with tunl0 address `172.16.166.0`), with a ping (`ping -c1 172.16.224.157`) onto the good pod, it looks like this:

```
$ tcpdump -i tunl0 -nn icmp
00:46:12.380880 ip: 172.16.166.0 > 172.16.224.157: ICMP echo request, id 28366, seq 1, length 64
00:46:12.381042 ip: 172.16.224.157 > 172.16.166.0: ICMP echo reply, id 28366, seq 1, length 64
```

With a ping (`ping -c1 172.16.224.147`) onto the bad pod, it looks like this (no response):

```
$ tcpdump -i tunl0 -nn icmp
00:47:38.756049 ip: 172.16.166.0 > 172.16.224.147: ICMP echo request, id 31423, seq 1, length 64
```

And here is the routing table on destination node for `172.16.166.0`:

```
$ ip r | grep 172.16.166
172.16.166.0/26 via 10.177.83.90 dev tunl0  proto bird onlink
```

So (1) the routing table looks good, and (2) it's slightly different with @MikeSpreitzer 's case: it either can ping a pod (172.16.224.157) successfully, or failed to ping another pod (172.16.224.147) **on the same node**.

And obviously 147 and 157 are in the same CIDR: `172.16.224.128/26 via 10.177.20.24 dev tunl0  proto bird onlink`

**Comment by Huang-Wei** [comment]
@fasaxc I found it has nothing to do with IPIP. It's a issue in `calico-policy-controller`. I opened https://github.com/projectcalico/k8s-policy/issues/72 to track this.

Thanks for your help.

**Comment by tmjd** [comment]
@Huang-Wei it sounds like you figured out the problem so I'm going to close this issue.  If I'm incorrect please feel free to reopen.

**Comment by tmjd** [comment]
I think I closed this prematurely before. It sounds like Mike has a work around for this issue though.  @fasaxc should we keep this issue open to track it or is there a more specific issue that can be created to track the root of the problem?

**Comment by fasaxc** [comment]
Looking at Mike's logs fromt he worker node and comparing the confd log and the bird log, I see that BIRD comes up and connects to 100.127.0.2 and finishes syncing routes:
```
2016-12-02_20:14:06.10481 bird: Mesh_100_127_0_2: Connected to table master
2016-12-02_20:14:06.10484 bird: Mesh_100_127_0_2: State changed to feed
2016-12-02_20:14:06.10484 bird: Mesh_100_127_0_2: State changed to up
2016-12-02_20:14:06.70085 bird: Mesh_100_127_0_3: Connected to table master
2016-12-02_20:14:06.70087 bird: Mesh_100_127_0_3: State changed to feed
2016-12-02_20:14:06.70088 bird: Mesh_100_127_0_3: State changed to up
2016-12-02_20:14:06.70281 bird: Mesh_100_127_0_6: Connected to table master
2016-12-02_20:14:06.70282 bird: Mesh_100_127_0_6: State changed to feed
2016-12-02_20:14:06.70283 bird: Mesh_100_127_0_6: State changed to up
2016-12-02_20:14:06.70350 bird: Mesh_100_127_0_4: Connected to table master
2016-12-02_20:14:06.70351 bird: Mesh_100_127_0_4: State changed to feed
2016-12-02_20:14:06.70352 bird: Mesh_100_127_0_4: State changed to up
2016-12-02_20:14:06.70352 bird: Graceful restart done
```

Then, I think, confd spots that IPIP is enabled:
```
2016-12-02T20:14:12Z mvespa1-5 confd[88]: INFO /tmp/tunl-ip has md5sum ec31acd5f31714c49514fcd2aa56e741 should be fa79d3c99b02b10c177820d456024e6a
2016-12-02T20:14:12Z mvespa1-5 confd[88]: INFO Target config /tmp/tunl-ip out of sync
2016-12-02T20:14:12Z mvespa1-5 confd[88]: DEBUG Overwriting target config /tmp/tunl-ip
2016-12-02T20:14:12Z mvespa1-5 confd[88]: DEBUG Running allocate-ipip-addr
2016-12-02T20:14:12Z mvespa1-5 confd[88]: DEBUG Comparing candidate config to /etc/calico/confd/config/bird_ipam.cfg
2016-12-02T20:14:12Z mvespa1-5 confd[88]: INFO /etc/calico/confd/config/bird_ipam.cfg has md5sum 8cec8ff0ddd9971c6e599981963051a8 should be 0d1b99638ee59053b25193416405b006
```
and SIGHUP's bird to reconfigure it:
```
2016-12-02_20:14:12.34261 bird: Reconfiguration requested by SIGHUP
2016-12-02_20:14:12.34279 bird: Reconfiguring
```
However, by that point, the "bad" route:
```
172.16.184.128/26 via 100.127.0.2 dev enp0s3  proto bird
```
must already by in the routing table, and BIRD doesn't fix it when it gets the SIGHUP.

**Comment by tomdee** [comment]
Should this be relabelled as a bug?

**Comment by fasaxc** [comment]
@MikeSpreitzer, after some sleuthing, we think this issue was fixed by the following commit, which was shipped in calico/node v1.0.0: https://github.com/projectcalico/bird/commit/205aa82fa23d05702ae11b4fa7fb09135e9955e1.  Are you still seeing the issue in v1.0.0+?

**Comment by caseydavenport** [comment]
Closing since we think this is fixed in recent releases.
